[DO NOT REVIEW] add watchdog canister

maksymar opened the pull request:
DO NOT REVIEW: going to split this PR to separate fetching #177 & serving data.

This PR adds a watchdog_canister for checking bitcoin_canister height health.

Current implementation does not use `ic-http` for testing HTTP outcalls, should be added later.

ielashi commented on watchdog/src/remote_api/bitcoin_canister.rs:
```diff
@@ -0,0 +1,123 @@
+impl BitcoinCanister {
+    /// The host of the remote API.
+    pub fn host() -> &'static str {
+        "ghsi2-tqaaa-aaaan-aaaca-cai.raw.ic0.app"
```
nits: I'd make the bitcoin canister ID part of the configuration of this canister, in case we want to run a similar watchdog for the bitcoin testnet canister, for example.

The `raw.ic0.app` isn't secure. Maybe we can add a note/todo to replace this with a more secure alternative in the future? We've received requests before to expose the height/tip block hash and other info as an endpoint in the bitcoin canister, so we could consider adding that in the future.

maksymar replied:
Done.

ielashi commented on watchdog/src/remote_api/blockstream_info.rs:
```diff
@@ -0,0 +1,68 @@
+use crate::types::BlockHeight;
+use ic_cdk::api::management_canister::http_request::CanisterHttpRequestArgument;
+
+pub struct BlockstreamInfo {}
```
nit: do we need this struct? Is there an advantage to having this struct vs just having the methods below as pure functions?

ielashi commented on watchdog/src/remote_api/api_bitaps_com.rs:
```diff
@@ -0,0 +1,98 @@
+use serde_json::json;
+
+/// The transform function for the remote API.
+#[ic_cdk_macros::query]
+fn transform_api_bitaps_com(raw: TransformArgs) -> HttpResponse {
+    apply_to_body_json(raw, ApiBitapsCom::transform)
+}
```
nit: (feel to ignore) I personally find it clearer if all of the canister's endpoints are in `main.rs`, so that it's clear to the reader what the canister exposes to the world.

ielashi commented on watchdog/src/time.rs:
```diff
@@ -0,0 +1,49 @@
+use std::time::Instant;
+
+#[cfg(not(target_arch = "wasm32"))]
+thread_local! {
+    /// Canister inner uptime since start in nanoseconds.
+    static START: RefCell<Instant> = RefCell::new(Instant::now());
+}
```
nit: similar to the comment about listing all the endpoints in `main.rs`, it's in general a good practice to keep all of the canister's state in one place, to make it easy for the reader to see. This is documented in the "Effective Rust Canisters" document: https://mmapped.blog/posts/01-effective-rust-canisters.html#clear-state

ielashi reviewed (approved):
LGTM, thanks @maksymar!

ielashi commented:
You'll want to fix the clippy issues for CI to proceed, or temporarily add `allow` statements in the code until the `ic-http-mock` code is merged.

maksymar commented on watchdog/src/info.rs:
```diff
@@ -0,0 +1,301 @@
+            None => Status::new(StatusCode::NoBitcoinCanisterData, None),
+            Some(bitcoin_canister_height) => {
+                if info.explorers_n < info.config.min_explorers {
+                    // No enough explorers.
```
Done.

maksymar commented:
> nit: do we need this struct? Is there an advantage to having this struct vs just having the methods below as pure functions?

> nit: (feel to ignore) I personally find it clearer if all of the canister's endpoints are in `main.rs`, so that it's clear to the reader what the canister exposes to the world.

> nit: similar to the comment about listing all the endpoints in main.rs, it's in general a good practice to keep all of the canister's state in one place, to make it easy for the reader to see. This is documented in the "Effective Rust Canisters" document: https://mmapped.blog/posts/01-effective-rust-canisters.html#clear-state

In order to address those comments I refactored the code a bit and split to make PRs smaller.
Here is the part that does all the HTTP requests and collects the responses. Processing and serving for bitcoin_canister will be done in separate PRs.

PTAL at #177
